docs: add proposal for CI without pull_request_target (#13333)
kata-containers/kata-containers: docs/design/proposals/ci-without-pull-request-target.md

!!! info
    The three `AZ_*` values are identifiers used with OIDC federated
    credentials rather than long-lived passwords, but they are stored as
    secrets and therefore carry the same fork-visibility restriction.

## Threat model and design constraints

What we are protecting against, in rough priority order:

1. **Secret exfiltration** — a fork PR reads `QUAY_DEPLOYER_PASSWORD`,
   `NGC_API_KEY`, `AUTHENTICATED_IMAGE_PASSWORD`, `KBUILD_SIGN_PIN`, or mints an
   Azure OIDC token, and leaks it.
2. **Registry poisoning** — a fork PR uses the write-scoped token to push a
   malicious image to `ghcr.io/kata-containers/*` that later runs on trusted
   infrastructure or is consumed by users.
3. **Infrastructure abuse** — a fork PR uses Azure credentials to spin up
   arbitrary resources (crypto-mining, lateral movement).

Design constraints that any alternative must respect:

- **Fork contributors must still get meaningful CI.** The build and the bulk of
  the tests should run for external PRs, ideally without a maintainer babysitting
  every push.
- **Multi-arch reality.** Builds and tests span `amd64`, `arm64`, `s390x`,

>   `ppc64le`, plus bare-metal self-hosted runners (NVIDIA GPU, zVSI). Not every

Reviewer: May we leave a TODO here: span on `riscv64` as well (not yet 🥲)

  alternative is available on every runner class.
- **Don't regress the release path.** SLSA provenance attestation runs only for
  `stage: release` (post-merge), not on fork PRs, so it is unaffected by this
  work — but changes to the shared build workflows must not break it.

## General strategies

The alternatives below draw from a small set of building blocks. Most per-usage
recommendations are combinations of these.

> !!! note "These compose; C is never standalone"

Reviewer 2: Sorry for the dumb question, but is there a way to see the mkdoc output of this in the PR as it doesn't seem to render well in github?

Reviewer 3: Jfyi: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

`nvidia` assets, to sign out-of-tree GPU driver kernel modules.

**Why it needs `pull_request_target`.** Static secret; the signing happens
inside the (untrusted) build.

**Alternatives.**

- **Preferred: don't sign on the PR path.** Use an ephemeral, throwaway signing
  key generated in-job for PR builds (modules load on the CI test kernel because
  the CI guest kernel trusts the ephemeral key, or Secure Boot enforcement is off
  for the test guest). The real `KBUILD_SIGN_PIN` is used only for
  release/nightly builds that run outside fork context. This removes the secret
  from fork-triggered runs entirely.
- **Alternative: strategy A** — build unsigned modules on the PR path, sign in a
  trusted `workflow_run` job. More complex; only worth it if PR tests genuinely
  require production-signed modules.

**Recommendation.** Ephemeral per-build signing key on the PR path; keep the
production PIN in release/nightly only.

### 4. Azure AKS provisioning (`AZ_APPID` / `AZ_TENANT_ID` / `AZ_SUBSCRIPTION_ID`)

**Today.** `azure/login` (OIDC, `id-token: write`) creates and tears down AKS
clusters for `run-k8s-tests-on-aks.yaml` and `run-kata-coco-tests.yaml`.

> **Why it needs `pull_request_target`.** The identifiers live in secrets, and the
> OIDC token is minted inside a workflow that also runs fork code.

Reviewer 3: Just to be clear, based on elsewhere in this proposal:
```suggestion
**Why it needs `pull_request_target`.** The identifiers live in secrets, and the `id-token: write` permission.
```

BTW, I used to think we could unblock `pull_request` for Azure OIDC by converting the identifiers from secrets to variables (if secure), but as I just learned from this proposal, that wouldn't be enough since `id-token: write` also requires `pull_request_target`.

**Alternatives.**

- **Preferred: environment-scoped OIDC (strategy B + C).** These are already
  OIDC, not passwords — the leverage is in **constraining the federated
  credential**. Bind the Azure federated credential to a protected `environment`
  (e.g. `aks-ci`) whose subject claim only matches runs approved to use it, and
  scope the Azure role to a dedicated, quota-limited resource group. A fork PR
  then cannot mint a usable token without approval, and even an approved token

>   can only touch the CI resource group.

Reviewer 3: I'd cautiously agree this would be preferred.

IIRC the federated credential is already tied to environment (`ci`), although I'm unsure what "protected" means here.

Re: the Azure role scope, I think it's currently rather permissive and I agree tightening it would be better.

- **Alternative: strategy A** — provision/own the cluster from a trusted job and
  hand only a kubeconfig (or run tests against a long-lived, isolated CI cluster)
  to the untrusted job. The untrusted job never holds Azure credentials, only

>   Kubernetes access to a sandboxed namespace/cluster.

Reviewer 3: Seems this is suggesting to either:

* Pass kubeconfig around: isn't this essentially reinventing the wheel to expose secrets? Seems unsafe?
* Set up new self-hosted runners: I don't think we want to open that can of worms given how much time the community already spends maintaining/fixing the existing self-hosted runners.

- **Alternative: move AKS tests off the fork-PR path** and run them

>   post-merge/nightly, keeping only free-runner k8s tests on PRs.

Reviewer 3: I don't think this is realistic without hurting quality.

kubeconfig from a trusted provisioning job.

### 5. NVIDIA GPU tests (`NGC_API_KEY`)

**Today.** `run-k8s-tests-on-nvidia-gpu.yaml` exports `NGC_API_KEY` to
`gha-run.sh run-nv-tests` on self-hosted `amd64-nvidia-a100` runners.

**Why it needs `pull_request_target`.** Static secret used inside the test job.

**Alternatives.**

- **Preferred: strategy D (runner-resident credential).** Because these run on
  trusted bare-metal self-hosted runners, provision `NGC_API_KEY` on the
  runner/host (env file, or a runner-scoped secret) so it never flows through a
  fork-triggered workflow. Gate fork PRs onto these runners behind approval.
- **Alternative: OIDC/short-lived NGC token** if NGC supports federation, minted
  per approved run.

**Recommendation.** Store on the trusted GPU runner; drop `NGC_API_KEY` from the
workflow secret set.

### 6. Authenticated-registry CoCo test (`AUTHENTICATED_IMAGE_PASSWORD`)

**Today.** `quay.io` credentials so CoCo tests can exercise pulling from an
authenticated registry (`run-kata-coco-tests*.yaml`,
`run-k8s-tests-on-zvsi.yaml`).

**Why it needs `pull_request_target`.** Static secret consumed by the test.

**Alternatives.**

> - **Preferred: pull-only, low-value robot account.** This secret only needs to

Reviewer 2: I think this is already the case - we created a separate robot account just for this repo that doesn't have permissions for any others.

  **pull** a fixed busybox-based test image. Replace it with a dedicated
  pull-only robot account whose scope is a single throwaway repo. A leak of a
  pull-only credential to a public test image is low impact, which makes it
  acceptable to expose even under the current model — or to inject via a
  protected environment (strategy B).

> - **Alternative: host a local registry in-cluster** seeded during the trusted

Reviewer 2: We did try this in peer pods about 3 years ago, but IIRC setting up a local registry with HTTPS was not something we managed to solve, so gave up. Maybe AI (or something smarter) could get this working now though

    end
    APPROVE --> T3a & T3b & T3c & T3d
  end

  RESULT["commit status / check-run on head_sha (API)"]:::trusted
  UT1 & UT2 & UC1 --> RESULT
  T2a & T2b & T2c --> RESULT
  T3a & T3b & T3c & T3d --> RESULT

  subgraph G["GATEKEEPER — on: pull_request_target — DEFAULT-BRANCH code — read-only — no secrets"]
    direction TB
    GK1["base-branch skips.py/jobs.py + PR diff as DATA"]:::trusted
    GK2["compute required jobs · poll statuses · post required check"]:::trusted
    GK1 --> GK2
  end
  RESULT -. statuses .-> GK2
  GK2 --> MERGE{{"branch protection: merge?"}}:::gate
```

!!! info "What is automatic vs. what needs a human"
    The `workflow_run` trigger is **fully automatic**: the Trusted lane starts on
    its own the instant the Untrusted build workflow completes — there is no
    manual dispatch, and it runs from the default branch with secrets + a write
    token regardless of the fork origin. It fires on **any** conclusion, so
    "did the build succeed" is a `workflow_run.conclusion == 'success'` check,
    not the trigger itself. The **only** manual step is optional and deliberate:
    any job attached to a protected **environment** with *required reviewers*

>     pauses until a maintainer approves it. That is where today's `ok-to-test`

Reviewer 2: This will probably come across as a massive nit, but what is your definition of a maintainer here - is it more than someone with committer access as we don't really have a proper process and definition of maintainers yet and I don't want to see this just being the responability of the usual minority if we can help it.

Author: Anyone who can currently add an `ok-to-test` label.

Reviewer 2: ok, I guess that it anyone with write access, so our committers list. Thanks!

    on the Untrusted `pull_request` side for *first-time contributors* — a
    platform default, unrelated to this design.

### Trust boundary: publish is data-only, test lanes are not

!!! danger "The test lanes still run PR code — the trigger alone does not fix that"
    It is tempting to say "the trusted lane never runs fork code", but that is
    only true of the **publish** step and the **gatekeeper**, which run
    default-branch code and treat every PR-derived input (`head_sha`, the diff,
    the downloaded artifact) as inert **data**. The **test lanes are
    different**: to test a PR they must deploy the PR-built image *and* run the
    PR's test harness (`tests/…/gha-run.sh`, bats, genpolicy testdata) on
    runners that hold credentials. That is unavoidable — testing untrusted code
    means running untrusted code near secrets. `workflow_run` does **not** make
    that safe; the guardrails for the test lanes are the **approval gate**
    (a human reviews the PR, including its test/workflow diff) plus
    **minimized, scoped, ephemeral credentials** (short-lived OIDC scoped to a
    quota-capped resource group, pull-only robot accounts, runner-resident keys,

>     disposable clusters/namespaces).

Reviewer 4: This is something I had in my head since I started reading this proposal and IMO belongs at the beginning as well. All we can do is to minimize the impact as in the end we will run the built content on the runners along with secrets. People should keep this in mind and not forget and assume this new design solves all security problems (I'm not saying people do that but it wouldn't hurt state it at the beginning and perhaps in the end as well).

The way I understand this proposal is it almost/completely mitigates the problem of injecting ugly shellscript that is used by our workflows but there are some exceptions like the bat files which need the tokens to perform the actions to test the new code. Similarly the kata-deploy on self-hosted runners, those remain the attack surface.

   read-only token) can stay; only move it to `workflow_run` + a label shim if a
   zero-`pull_request_target` policy is desired.

## Risks and open questions

- **`workflow_run` ergonomics.** Two-stage pipelines complicate status
  reporting and required-check configuration. If the gatekeeper stays on
  `pull_request_target` (the recommended path) it must still learn to correlate
  the CI's `workflow_run`-based jobs; if it is moved to `workflow_run`, it also
  needs a label-events shim for `ok-to-test` / `required-labels`.
- **`workflow_run` artifact hand-off.** The build (untrusted) and push (trusted)
  halves must agree on artifact names and the image tag scheme, and the trusted
  half must resolve the PR head SHA from the trigger metadata rather than trust
  any attacker-supplied input.
- **Which secrets are truly exercised on the PR path vs only nightly/release?**
  This should be measured (not assumed) before removing any secret from
  `ci.yaml`, to avoid silently disabling coverage.
- **Environment approval fatigue.** Protected environments re-introduce a manual
  step; the project should decide which lanes require approval for forks vs. run

>   freely.

Reviewer 4: Another question is the number of workflows, which used to be a problem in this repo. IIUC we'll have to create a couple more to separate the phases. Does anyone knows if the limit increased? It was about 23 workflows last time I suffered the consequences.

Reviewer 2: We got our special agent on the inside to increase our limit, so I think we can have 50 reusable workflows called now.

Reviewer 3: 🕵️‍♂️

GH Support claims we can use 100 concurrent workflows now but we haven't put that limit to the test I believe.
